File: Array/Easy/arrayUnion.py
#make array uinon of two sorted array
def unionArray(arr1,arr2):
    temp = set()
    res=[]
    for i in range(0,len(arr1)):
        temp.add(arr1[i])
    
    for i in range(0,len(arr2)):
        temp.add(arr2[i])
    
    for i in temp:
        res.append(i)
    return res

# arr1=[1,2,3,4,5]
# arr2=[5,6,7]
# print(unionArray(arr1,arr2))

def unionOptimize(arr1,arr2):
    i,j=0,0
    union=[]
    while i < len(arr1) and j < len(arr2):
        if arr1[i] <= arr2[j]:
            if len(union) == 0 or union[-1] != arr1[i]:
                union.append(arr1[i])
            i+=1
        else :
            if len(union) == 0 or union[-1] != arr2[j]:
                union.append(arr2[j])
            j+=1
    
    while i < len(arr1):
        if union[-1] != arr1[i]:
            union.append(arr1[i])
        i+=1

    while j < len(arr2):
        if union[-1] != arr2[j]:
            union.append(arr2[j])
        j+=1
    return union
   
        
arr1=[1,2,3,4,5]
arr2=[5,6,7]
print(unionOptimize(arr1,arr2))

File: Array/Easy/js/arrayUnion.js
/** Array Union */

let arr1=[1,2,3,4,5]
let arr2=[5,6,7]

function unionArray(arr1, arr2){
    for (let i=0;i<arr2.length;i++){
        if (!arr1.includes(arr2[i])){
            arr1.push(arr2[i])
        }
    }
    return arr1;
}

/** Using set DS */
function unionSet(a,b){
    let set = new Set();
    for(let val of a){
        set.add(val)
    }
    for(let val of b){
        set.add(val)
    }

    // let itr = set.entries();
    // for(let i=0;i<set.size();i++){

    // }
    console.log(set)
}

/** Using 2 pointer approach */
function union2P(a, b){
    let N1 = a.length;
    let N2 = b.length;
    let i = 0;
    let j=0;
    let union = []
    while(i<N1 && j<N2){
        if(a[i] <= b[j]){
            if(union.length == 0 || union[union.length-1] != a[i]) {
                union.push(a[i])
            }
            i++;
        } else if (a[i] >= b[j]){
            if(union.length == 0 || union[union.length-1] != b[j]) {
                union.push(b[j])
            }
            j++;
        }        
    }

    while(i<N1){
        if(union.length == 0 || union[union.length-1] != a[i]) {
            union.push(a[i])
        }
        i++;
    }

    while(j<N2){
        if(union.length == 0 || union[union.length-1] != b[i]) {
            union.push(b[i])
        }
        j++;
    }
    console.log(union)
}

console.log(unionArray(arr1,arr2))